docs/proposals/synchronous-garbage-collection.md

```diff
@@ -0,0 +1,126 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Overview
+
+Some users of the server-side garbage collection need to tell if the garbage collection is done ([example](https://github.com/kubernetes/kubernetes/issues/19701#issuecomment-236997077)). Synchronous Garbage Collection is a best-effort (see [unhandled cases](#unhandled-cases)) mechanism to enable such use cases: after the API server receives a deletion request of an owning object, the object keeps existing in the key-value store until all its dependents are deleted from the key-value store by the garbage collector.
+
+Tracking issue: https://github.com/kubernetes/kubernetes/issues/29891
+
+# Required code modification
+
+We need to make changes in the API, the API Server, and the garbage collector to support synchronous garbage collection.
+
+## API changes
+
+```go
+DeleteOptions {
+  …
+  // If SynchronousGarbageCollection is set, the object will not be deleted immediately. Instead, a GarbageCollectionInProgress finalizer will be placed on the object. The garbage collector will remove the finalizer from the object when all depdendents are deleted.
+  // SynchronousGarbageCollection and OrphanDependents are exclusive.
+  // SynchronousGarbageCollection default to false.
+  // SynchronousGarbageCollection is cascading, i.e., the object’s dependents will be deleted with the same SynchronousGarbageCollection.
+  SynchronousGarbageCollection *bool
+}
+```
+
+We will introduce a new standard finalizer: const GCFinalizer string = “GarbageCollectionInProgress”
+
+## Components changes
+
+### API Server
+
+Delete() function needs to check the DeleteOptions.SynchronousGarbageCollection.
+
+* The option is ignored if DeleteOptions.OrphanDependents is true or nil.
+* If the option is set, the API server will update the object instead of deleting it, add the finalizer, and set the `ObjectMeta.DeletionTimestamp`.
+
+### Garbage Collector
+
+**Modifications to processEvent()**
+
+`processEvent()` manages GC’s internal owner-dependency relationship graph, `uidToNode`. It updates `uidToNode` according to the Add/Update/Delete events in the cluster. To support synchronous GC, it has to:
+
+* handle Add or Update events where `obj.Finalizers.Has(GCFinalizer) && obj.DeletionTimestamp != nil`. The object will be added into the `synchronousGC queue`. The object will be marked as “GC in progress” in `uidToNode`.
```

Comment:
We can. When processing the queue, GC needs to check what kind of item it is, and apply different operations. I think that will make the code hard to reason, so I prefer using different queue. Are you worrying about performance? Or are you worrying about synchronization problem?